packages/twenty-server/src/engine/middlewares/middleware.service.ts

```diff
@@ -0,0 +1,98 @@
+import { Injectable } from '@nestjs/common';
+
+import { Request, Response } from 'express';
+import { ExtractJwt } from 'passport-jwt';
+
+import { AccessTokenService } from 'src/engine/core-modules/auth/token/services/access-token.service';
+import { DataSourceService } from 'src/engine/metadata-modules/data-source/data-source.service';
+import { WorkspaceMetadataCacheService } from 'src/engine/metadata-modules/workspace-metadata-cache/services/workspace-metadata-cache.service';
+import { EXCLUDED_MIDDLEWARE_OPERATIONS } from 'src/engine/middlewares/constants/excluded-middleware-operations.constant';
+import { bindDataToRequestObject } from 'src/engine/middlewares/utils/bind-data-to-request-object.utils';
+import { WorkspaceCacheStorageService } from 'src/engine/workspace-cache-storage/workspace-cache-storage.service';
+import { isDefined } from 'src/utils/is-defined';
+
+@Injectable()
+export class MiddlewareService {
+  constructor(
+    private readonly accessTokenService: AccessTokenService,
+    private readonly workspaceStorageCacheService: WorkspaceCacheStorageService,
+    private readonly workspaceMetadataCacheService: WorkspaceMetadataCacheService,
+    private readonly dataSourceService: DataSourceService,
+  ) {}
+
+  public excludedOperations = EXCLUDED_MIDDLEWARE_OPERATIONS;
+
+  public isTokenPresent(request: Request): boolean {
+    const token = ExtractJwt.fromAuthHeaderAsBearerToken()(request);
+
+    return !!token;
+  }
+
+  public checkUnauthenticatedAccess(request: Request): boolean {
+    const { body } = request;
+
+    const isUserUnauthenticated = !this.isTokenPresent(request);
+    const isExcludedOperation =
+      !body?.operationName ||
+      this.excludedOperations.includes(body.operationName);
+
+    return isUserUnauthenticated && isExcludedOperation;
+  }
+
+  public writeResponseOnExceptionCaught(
+    res: Response,
+    source: 'rest' | 'graphql',
+    error: unknown,
+    errors: unknown[],
+  ) {
+    const getStatus = () => {
+      if (source === 'graphql') {
+        return 200;
+      }
+
+      if (this.hasErrorStatus(error)) {
+        return error.status;
+      }
+
+      return 500;
+    };
+
+    res.writeHead(getStatus(), { 'Content-Type': 'application/json' });
+    res.write(
+      JSON.stringify({
+        errors,
+      }),
+    );
+    res.end();
+  }
```

Comment:
I think we should split that into 2 methods and remove the `source` parameter

```suggestion
  public writeGraphqlResponseOnExceptionCaught(
res: Response,
    error: unknown,
    errors: unknown[],
) {...    

  public writeRestResponseOnExceptionCaught(...
```